minio/minio
# Add support for ILM transition.

This PR adds transition support for ILM
to transition data to another MinIO target
represented by a storage class ARN. Subsequent
GET or HEAD for that object will be streamed from
the transition tier. If PostRestoreObject API is
invoked, the transitioned object can be restored for
duration specified to the source cluster.

## Description
Needs https://github.com/minio/minio-go/pull/1390 and https://github.com/minio/mc/pull/3420

## Motivation and Context
Allows applications to transition data to less expensive storage, e.g from NVME to hard disks. 

## How to test this PR?
Use companion PR(s) listed above. 
```
1. Set up a transition target for srcbucket on myminio to dstbucket on minio2:9000
mc admin bucket remote add  myminio/srcbucket \
        https://foobar:foo12345@minio2:9000/dstbucket \
        --service "ilm" --region "us-west-1" --label "transitiontier"

2. Use arn generated above to add a ilm rule with transition destination specified by the --storageclass flag
mc ilm add --expiry-days 3 --transition-days 2 --storage-class "transitiontier" source/srcbucket                         

When a object qualifies for transitioning per the rules set in the ilm config, it should be moved to the destination cluster. However all GET and HEAD should work. If post restore object API is called on a transitioned object, it should generate a local copy of object on source cluster.

For example:
aws s3api restore-object --endpoint-url http://localhost:9008 --profile minio \
--bucket srcbucket \
--key x2 \
--restore-request Days=1
```

## Types of changes

## Checklist:
